Improve error handling and rename go project

DaniElectra opened the pull request:
Now that we can return errors to nex-protocols-go and it will handle it,
we can improve how errors are being handled in friends.

There's probably room for improvement from this, since I haven't added
error handling on some database getters due to uncertainties on whether
the error should interrupt an action or not.

Although the servers are confirmed to be working, it hasn't been tested
extensively, so the PR should be reviewed carefully.

The MongoDB client has also been removed, in favor of GRPC calls to the
account server (using `GetUserData` specifically), and the README style
and Makefile from Pikmin 3 has been imported after renaming.

The Go project has been renamed from `friends-secure` to `friends`, to
keep consistency with the GitHub name.

DaniElectra committed:
Improve error handling

Now that we can return errors to nex-protocols-go and it will handle it,
we can improve how errors are being handled in friends.

There's probably room for improvement from this, since I haven't added
error handling on some database getters due to uncertainties on whether
the error should interrupt an action or not.

DaniElectra committed:
Rename friends-secure to friends and remove Mongo

Replace MongoDB usage with GRPC calls to the account server. Also
include documentation on the README and a Makefile.

jonbarrow commented on database/3ds/get_friend_miis.go:
```diff
@@ -4,13 +4,13 @@ import (
 
-// Get a friend's mii
+// GetFriendMiis returns the Mii of all friends
 func GetFriendMiis(pids []uint32) []*friends_3ds_types.FriendMii {
```
Since we're here, this should also return an error. The logging and whether or not the request should respond with an error should be done in the request handler honestly, putting the logging in this functions doesn't make a _ton_ of sense

jonbarrow commented on database/3ds/get_friend_persistent_infos.go:
```diff
@@ -4,13 +4,13 @@ import (
 
-// Get a friend's persistent information
+// GetFriendPersistentInfos returns the persistent information of all friends
 func GetFriendPersistentInfos(user1_pid uint32, pids []uint32) []*friends_3ds_types.FriendPersistentInfo {
```
Same comment about this returning an error

jonbarrow commented on database/3ds/get_user_friends.go:
```diff
@@ -3,12 +3,12 @@ package database_3ds
 
-// Get all of a user's friend relationships
+// GetUserFriends returns all friend relationships of a user
 func GetUserFriends(pid uint32) []*friends_3ds_types.FriendRelationship {
```
Same comment about this returning an error

jonbarrow commented on database/wiiu/accept_friend_request_and_return_friend_info.go:
```diff
@@ -3,21 +3,19 @@ package database_wiiu
 )
 
-func AcceptFriendRequestAndReturnFriendInfo(friendRequestID uint64) *friends_wiiu_types.FriendInfo {
+func AcceptFriendRequestAndReturnFriendInfo(friendRequestID uint64) (*friends_wiiu_types.FriendInfo, error) {
```
Not your fault, but since we're here. This exported function is missing a comment

This happens elsewhere too but I don't want to spam this PR with tons of duplicate reviews

jonbarrow commented on database/wiiu/accept_friend_request_and_return_friend_info.go:
```diff
@@ -97,11 +96,8 @@ func AcceptFriendRequestAndReturnFriendInfo(friendRequestID uint64) *friends_wii
-			} else {
-				globals.Logger.Critical(err.Error())
-			}
+			// TODO: Should we return the error here?
```
Yes this should return an error. I can't remember why this is defaulted like this, but now whenever a user connects their user data is updated in `RegisterEx`. The only time this should ever error is if the data can't be found for the user in the friend request, which is critical tbh

jonbarrow commented on database/wiiu/get_user_info_by_pid.go:
```diff
@@ -1,44 +1,36 @@
+	info := friends_wiiu_types.NewPrincipalBasicInfo()
 
-	err := database.MongoCollection.FindOne(context.TODO(), bson.D{{Key: "pid", Value: pid}}, options.FindOne()).Decode(&result)
+	userData, err := globals.GetUserData(pid)
```
Since this uses grpc now, this really shouldn't be in the database package

It should also return an error

DaniElectra replied:
> Since this uses grpc now, this really shouldn't be in the database package

Not sure what you mean here?

jonbarrow replied:
The database packages are for functions which directly interact with the database

`GetUserData` is a gRPC call, this no longer interacts with the database directly

jonbarrow commented on database/wiiu/remove_friendship.go:
```diff
@@ -1,21 +1,22 @@
 
 // Remove a user's friend relationship
-func RemoveFriendship(user1_pid uint32, user2_pid uint32) {
+func RemoveFriendship(user1_pid uint32, user2_pid uint32) error {
```
Again not your fault, but since we're here. This has a comment but it's in the wrong format. Also happens elsewhere

jonbarrow commented on grpc/deny_friend_request.go:
```diff
@@ -3,15 +3,14 @@ package grpc
 func (s *gRPCFriendsServer) DenyFriendRequest(ctx context.Context, in *pb.DenyFriendRequestRequest) (*pb.DenyFriendRequestResponse, error) {
-	// TODO - Make this return an error
-	database_wiiu.SetFriendRequestDenied(in.GetFriendRequestId())
+	err := database_wiiu.SetFriendRequestDenied(in.GetFriendRequestId())
```
This should probably have a more robust error check, similar to `AcceptFriendRequest`

jonbarrow commented on grpc/send_user_friend_request.go:
```diff
@@ -24,9 +24,9 @@ func (s *gRPCFriendsServer) SendUserFriendRequest(ctx context.Context, in *pb.Se
 	message := in.GetMessage()
 
-	id := database_wiiu.SaveFriendRequest(sender, recipient, sentTime.Value(), expireTime.Value(), message)
+	id, _ := database_wiiu.SaveFriendRequest(sender, recipient, sentTime.Value(), expireTime.Value(), message)
```
This should probably have a more robust error check, similar to `AcceptFriendRequest`

jonbarrow commented on nex/account-management/nintendo_create_account.go:
```diff
@@ -8,64 +8,76 @@ import (
 		tokenBase64 = strings.Replace(tokenBase64, "*", "=", -1)
+	default:
+		globals.Logger.Errorf("Invalid oAuthData data type %s!", oAuthDataType)
+		return nex.Errors.Authentication.TokenParseError
```
Is this the real error which is sent in these cases?

DaniElectra replied:
Calling this method with dummy data, including replacing the oAuthData with friends.Mii raises `PythonCore.Exception`, although I think we are fine by using this error code

jonbarrow replied:
Ideally I would like to be ass accurate as possible, even when it comes to errors

Though that error sounds...different. It sounds like the real servers aren't actually checking the DataHolder type, which leads to an exception? So maybe you're right that we should use a different error, because that's not *actually* what's happening here for us, there is no exception

DaniElectra replied:
Should I keep this error then?

jonbarrow replied:
Yeah this is fine unless we find a better one in the future

jonbarrow commented on nex/friends-3ds/get_pid_by_lfc.go:
```diff
@@ -26,4 +26,6 @@ func GetPrincipalIDByLocalFriendCode(err error, client *nex.Client, callID uint3
 
 	globals.SecureServer.Send(responsePacket)
+
+	return 0
```
Since this is stubbed right now we can just remove this entire response and return `nex.Errors.Core.NotImplemented`, keeping the comment

jonbarrow commented on nex/friends-3ds/remove_friend_by_lfc.go:
```diff
@@ -26,4 +26,6 @@ func RemoveFriendByLocalFriendCode(err error, client *nex.Client, callID uint32,
 
 	globals.SecureServer.Send(responsePacket)
+
+	return 0
```
Also stubbed, can just return `nex.Errors.Core.NotImplemented`

jonbarrow commented on nex/friends-wiiu/mark_friend_requests_as_received.go:
```diff
@@ -1,16 +1,30 @@
+		if err != nil {
+			globals.Logger.Critical(err.Error())
+			markErr = err
+		}
```
I'm not sold on the idea that we should continue processing data when an error occurs. Though I'm also not really sure of a better way to handle this. Thoughts?

DaniElectra replied:
I had the same doubts, and I'm not sure of another way to handle it. This issue will probably appear on some database getters too, so we have to think about it.

I will bring up @hauntii for suggestions too

hauntii replied:
I think we should abort processing here, just because if the console gets an error (which it will no matter if we abort or not) it's just going to try again anyways, so there's no point in continuing if it's going to try and remark them anyways.

For the others that retrieve user lists in the others mentioned, I also don't think we should keep going since we don't want the user to get an incomplete list of their friends / blocked and think that they've been unfriended / their block has inexplicably disappeared.

DaniElectra replied:
That makes sense, thank you!

DaniElectra committed:
Add missing error handling

Now database getters do error handling aswell, and in case of any error,
the NEX or GRPC method throws a proper error about it.

There are still some doubts on how to handle errors when a list of data
is processed, so for now skip that element on the list and continue.

Also add some constant errors to do better error handling.

DaniElectra committed:
Replace stubs with NotImplemented

DaniElectra committed:
Add missing error handling on GRPC method

DaniElectra commented:
I've added the missing error handling for database getters. In order to distinguish error types easier, I also added some error constants to avoid collisions with `sql.ErrNoRows` errors.

We still have the issue about continue processing data when an error occurs on some cases, so I'll wait for more opinions on that

jonbarrow commented on database/3ds/get_friend_persistent_infos.go:
```diff
@@ -44,6 +40,7 @@ func GetFriendPersistentInfos(user1_pid uint32, pids []uint32) []*friends_3ds_ty
 				friendedAtTime = uint64(time.Now().Unix())
 			} else {
 				globals.Logger.Critical(err.Error())
+				continue
```
This is another one where I'm not sure if we should continue to process data if an error occurs. Tagging @hauntii here as well

jonbarrow commented on database/wiiu/get_user_block_list.go:
```diff
@@ -24,9 +29,21 @@ func GetUserBlockList(pid uint32) []*friends_wiiu_types.BlacklistedPrincipal {
+		if err != nil {
+			globals.Logger.Critical(err.Error())
+			continue
+		}
```
Another should we continue. @hauntii

jonbarrow commented on database/wiiu/get_user_friend_list.go:
```diff
@@ -52,8 +55,21 @@ func GetUserFriendList(pid uint32) []*friends_wiiu_types.FriendInfo {
+			if err != nil {
+				globals.Logger.Critical(err.Error())
+				continue
+			}
```
I know this will create several duplicate review comments but I'm mostly doing this to bring attention to all the areas, and to tag @hauntii for them all

jonbarrow commented on database/wiiu/get_user_friend_list.go:
```diff
@@ -77,27 +93,29 @@ func GetUserFriendList(pid uint32) []*friends_wiiu_types.FriendInfo {
-					globals.Logger.Critical(err.Error())
-				}
+				globals.Logger.Critical(err.Error())
+				continue
```
Beep boop @hauntii

jonbarrow commented on database/wiiu/get_user_info_by_pnid_data.go:
```diff
@@ -0,0 +1,33 @@
+	info.Mii.Datetime = nex.NewDateTime(0)
+
+	return info, nil
+}
```
This was marked as resolved before but it was not? The functions inside the `database` packages are functions with interact directly with the database. This function does not do any database calls, it formats a gRPC response into another type, so it should be moved to somewhere like `globals` or `utility` or something

DaniElectra replied:
Ohhh, I completely misunderstood this one 😅

jonbarrow reviewed (changes requested):
I said I was going to leave a comment for all of the "should we continue processing data" parts but then I got lazy

DaniElectra committed:
Move GetUserDataByPID to utility

DaniElectra committed:
Add missing error handling and update modules

DaniElectra committed:
Add missing error handling (part 2)

jonbarrow commented on database/3ds/get_friend_persistent_infos.go:
```diff
@@ -41,9 +35,9 @@ func GetFriendPersistentInfos(user1_pid uint32, pids []uint32) []*friends_3ds_ty
 		if err != nil {
 			if err == sql.ErrNoRows {
-				friendedAtTime = uint64(time.Now().Unix())
+				return make([]*friends_3ds_types.FriendPersistentInfo, 0), database.ErrFriendshipNotFound
```
Since this returns an error, the first return value can just be `nil`. Returning an initilized slice like this can give the impression that the return value can still be used, when it should not. It's also _slightly_ more performant since not calling `make` won't run all the initialization

This happens elsewhere

DaniElectra committed:
Replace return empty lists with nil
